### PR TITLE
BUG: Error when specifying int index containing NaN

### PR DESCRIPTION
**Pull request by gfyoung**
Finally got around to patching this.  Partially addresses #15187.

### CODE REVIEW

File: pandas/indexes/base.py [hunk 1]
```diff
@@ -585,7 +590,14 @@ def where(self, cond, other=None):
         if other is None:
             other = self._na_value
         values = np.where(cond, self.values, other)
-        return self._shallow_copy_with_infer(values, dtype=self.dtype)
+
+        dtype = self.dtype
+        if self._is_numeric_dtype and np.any(isnull(values)):
+            # We can't coerce to the numeric dtype of "self" (unless
```
Comment (jreback) [inline pandas/indexes/base.py]: -> ``isnull(values).any()``

Comment (gfyoung) [inline pandas/indexes/base.py]: Breaks if `values` is a scalar.

**Comment by jreback** [comment]
looks good. maybe some test cases for the fail conditions that you are now catching?

File: pandas/indexes/base.py (hunk 1 above)
Comment (jreback) [inline pandas/indexes/base.py]: hmm, that's annoying. ok then.

**Comment by gfyoung** [comment]
> maybe some test cases for the fail conditions that you are now catching?

Ah, good catch.  I thought I had added some.  Will do.

**Comment by gfyoung** [comment]
@jreback : Added tests, and everything is green.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**